### PR TITLE
Add `newline-after-import` rule

### PR DESCRIPTION
**Pull request by singles**
# newline-after-import (#244)

Reports if there's no new line after last import in group.

## Rule Details

Valid:

```js
import defaultExport from './foo'

const FOO = 'BAR'
```

```js
import defaultExport from './foo'
import { bar }  from 'bar-lib'

const FOO = 'BAR'
```

...whereas here imports will be reported:

```js
import * as foo from 'foo'
const FOO = 'BAR'
```

```js
import * as foo from 'foo'
const FOO = 'BAR'

import { bar }  from 'bar-lib'
```

## When Not To Use It

If you like to visually group module imports with its usage, you don't want to use this rule.

### CODE REVIEW

**Comment by jfmengels** [comment]
Looks good to me, but is it possible to also support `require()` at the top of the document? Makes it quite a bit more complicated, but I know I (and the one who requested #244, cc @sindresorhus) would like to have that.

**Comment by singles** [comment]
@jfmengels You mean, you would like it to work for `import foo from 'foo'` and/or `var foo = require('foo')`?

**Comment by sindresorhus** [comment]
@singles Yes :)

File: README.md [hunk 1]
```diff
@@ -48,6 +48,7 @@ This plugin intends to support linting of ES2015+ (ES6+) import/export syntax, a
 * Ensure all imports appear before other statements ([`imports-first`])
 * Report repeated import of the same module in multiple places ([`no-duplicates`])
 * Report namespace imports ([`no-namespace`])
+* Enforce new line after import not followed by another import([`newline-after-import`])
```
Comment (sindresorhus) [inline README.md]: Maybe:

```
Enforce a newline after import statements ([`newline-after-import`])
```

**Comment by singles** [comment]
@sindresorhus @jfmengels thank you guys for feedback. 

Regarding `require`. I didn't add support for `require` after reading docs, which says that `This plugin intends to support linting of ES2015+ (ES6+) import/export syntax`. 

Also I checked `test` folder for usage of `require` and none of the current rules support that, while you could think that for example `imports-first` should do that, because `require` is still an import, just other syntax. 

Please don't get me wrong - I'm not trying to be cocky or something, it just matter of consistency for me :)

**Comment by jfmengels** [comment]
You're right that this project doesn't support `require`. There are some rules coming in from another [project](https://github.com/jfmengels/eslint-plugin-import-order) which has rules that does support `require` statements.

From what I understood, this plugin supports rules that can infer something from static analysis using `import`. Using `require`, that becomes harder/not possible, and therefore, most rules can't be applied to uses of `require` (might be wrong, haven't had the chance to look at this plugin in too much detail). In this case (and the rule I mentioned above), it is simply a stylistic issue (which are new to the project too), and can be applied to both "methods".

@benmosher Thoughts on supporting `require` where possible (and first of all, here)?

File: README.md (hunk 1 above)
Comment (singles) [inline README.md]: Thx, fixed :)

**Comment by benmosher** [comment]
FWIW, `no-unresolved` does support both CommonJS and AMD module syntax, if you are using Browserify or Webpack. Could theoretically support other AMD module loaders if the resolvers were implemented.

For that rule, you may supply an options object like `{ commonjs: true, amd: true }` to enable one or both.

I think it makes sense to support CommonJS to the extent that 

- [x] the community wants it 
- [x] it's technically feasible

I think that is true on both counts in this case.

**Comment by singles** [comment]
Now when #247 is here and based on comments from #246 I think it would be nice to solve this as a part of #247. Just additional option called `newline-after-import-group`, which will give you empty line before any other code, even if you will create one group with all import types. 

What do you think?

**Comment by benmosher** [comment]
Tough call. I might enable this rule without worrying about ordering within the imports themselves, so I'm not sure it makes sense rolled into `order`. (@sindresorhus convinced me with a similar argument in discussion on #253)

I think this makes sense as you've implemented it for `import`s, and I think the `require` version would just ensure that the next line is either another `require` call (possibly an assignment, or could be side-effects only) or a blank line, i.e.

```js
const x = require('x')
require('jqueryui')
var y = require('y')

// ^ line above is enforced
function boringName(someArg) {
  return new Error("the worst function ever") 
}

let z = require('ugh-mutable-module-reference-ftl')

// ^ another enforced line here
function stahp() {
  process.exit(-∞)
}
```

No need to enforce on `require` calls except at module scope†, so you could implement with a `Program` node visitor with a loop over `program.body`.

†: unless @jfmengels or @sindresorhus say otherwise. I don't think it's important for v1 but I'm not going to use it, so what do _I_ know 😅 

I guess you could provide the same visitor for both `Program` and `BlockStatement` nodes, they each have a `body` field that is an array of `Statement`s.

**Comment by singles** [comment]
OK, rebased onto `master` and added support for `require` :)

File: src/rules/newline-after-import.js [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+      const nextToken = context.getSourceCode(node).getTokenAfter(node)
+
+      if (!nextToken) {
+        return
+      }
+
+      ensureNoForbiddenKeyword(context, node, nextToken, 'import')
+    },
+    CallExpression: function(node) {
+      if (node.callee.name === "require") {
```
Comment (jfmengels) [inline src/rules/newline-after-import.js]: Do you mind using single-quotes here? For consistency with the other strings.

File: src/rules/newline-after-import.js [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+        const nextTokens = context.getSourceCode(node).getTokensAfter(node, 2)
+        const tokenToInspect = nextTokens.length > 1 && nextTokens[0].type === 'Punctuator'
+          ? nextTokens[1]
+          : nextTokens[0]
+
+        if (!tokenToInspect) {
+          return
+        }
```
Comment (jfmengels) [inline src/rules/newline-after-import.js]: You could move this check to `ensureNoForbiddenKeyword` for both `import` and `require` cases.

File: src/rules/newline-after-import.js [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+  return nextToken.loc.start.line - node.loc.start.line
+}
+
+function ensureNoForbiddenKeyword(context, node, tokenToInspect, tokenValue) {
+  if (getLineDifference(node, tokenToInspect) === 1
+    && tokenToInspect.type === 'Keyword' && tokenToInspect.value !== tokenValue)
+  {
+    context.report({
+      loc: tokenToInspect.loc.start,
+      message: 'Expected empty line after import/require ' +
```
Comment (jfmengels) [inline src/rules/newline-after-import.js]: Since you know whether it's a `require` or an `import`, might as well indicate which one it is in the error message.

File: src/rules/newline-after-import.js [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+// Rule Definition
+//------------------------------------------------------------------------------
+
+function getLineDifference(node, nextToken) {
+  return nextToken.loc.start.line - node.loc.start.line
+}
+
+function ensureNoForbiddenKeyword(context, node, tokenToInspect, tokenValue) {
+  if (getLineDifference(node, tokenToInspect) === 1
+    && tokenToInspect.type === 'Keyword' && tokenToInspect.value !== tokenValue)
```
Comment (jfmengels) [inline src/rules/newline-after-import.js]: My bad, but I have a hard time understanding the name `ensureNoForbiddenKeyword` (never did any rules that dealt with tokens), and how/what exactly it it testing (except the line difference thing, that I got). Do you mind explaining it to me?

File: src/rules/newline-after-import.js (hunk 2 above)
Comment (singles) [inline src/rules/newline-after-import.js]: No problem, also see #280 :)

File: src/rules/newline-after-import.js (hunk 3 above)
Comment (singles) [inline src/rules/newline-after-import.js]: Good catch, done.

File: src/rules/newline-after-import.js (hunk 4 above)
Comment (singles) [inline src/rules/newline-after-import.js]: Yep, that would be useful, fixed.

File: src/rules/newline-after-import.js (hunk 5 above)
Comment (singles) [inline src/rules/newline-after-import.js]: Role of that function is to check line difference AND if another token is something different that `import/require` - if it is different, then report an error, because only thing that can be directly in next line is either `import` or `require`.

**Comment by jfmengels** [comment]
LGTM :)

**Comment by singles** [comment]
@benmosher no problem, done.

**Comment by benmosher** [comment]
Rebased and added changelog updates. 👍  it's in master, ready for publish via 1.7.0. Not sure when that will be, probably want to get some of @sindresorhus's issues resolved first.

**Comment by jfmengels** [comment]
@benmosher It's not very nice of you to say that @sindresorhus has issues :p

**Comment by sindresorhus** [comment]
It's true. I have a [gadzillion issues](https://github.com/search?q=user%3Asindresorhus&state=open&type=Issues&utf8=%E2%9C%93)...

**Comment by jfmengels** [comment]
You made me found out I had way more issues than I thought... If it makes you feel better, I have a 3.5 hour trip, and I'll spend at least some of it solving your issues ^^
